[quant] QNNPACK deconv kernel and tests

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37497 [quant] QNNPACK factoring out common structs from conv/deconv
* **#36790 [quant] QNNPACK deconv kernel and tests**
* #37405 [quant] QNNPACK deconvolution packing



Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -19,7 +83,242 @@ enum pytorch_qnnp_status qnnpackDeConv(
+
+  // Check all unsupported parameters
+  const float deconvolution_scale = input_scale * kernel_scale / output_scale;
+  if (deconvolution_scale >= 1.0f) {
```
@kimishpatel does the float requantization PR apply to the deconv operation too?

kimishpatel replied:
Yes it does. And it is covered under conv operator. Please see this PR: https://github.com/pytorch/pytorch/pull/35856. I suggest we wait for it to land? Else I will have to make fixes here.

z-a-f replied:
Adding a TODO in here in case this lands before #35856

kimishpatel replied:
#35856 has landed.

z-a-f replied:
Can you elaborate what changes the PR will require?

kimishpatel replied:
I was gonna say like this: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc#L308. But it seems that I missed removing this check since it is not longer a requirement. @supriyar, did this not trigger on the model we ran the PR for requant changes? Seems strange.
@z-a-f I think you can remove this check.

supriyar replied:
>> @supriyar, did this not trigger on the model we ran the PR for requant changes? Seems strange.

I had made the changes locally when I ran the model so wasn't an issue.

z-a-f replied:
@kimishpatel I am confused, we don't need this condition anymore? Can I just remove the if statement defined in line 156 completely?

kimishpatel replied:
Yes, we dont need this condition anymore. Please remove it. I need to remove similar ones from conv-run and fc-run.

z-a-f replied:
Will do before landing

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -19,7 +83,242 @@ enum pytorch_qnnp_status qnnpackDeConv(
+  pytorch_qnnp_operator_t deconvolution =
+      static_cast<pytorch_qnnp_operator_t>(
+          calloc(1, sizeof(struct pytorch_qnnp_operator)));
+  if (deconvolution == NULL) {
```
Can we use nullptr instead of NULL in this file?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -19,7 +83,242 @@ enum pytorch_qnnp_status qnnpackDeConv(
+    }
+    default:
+      pytorch_qnnp_log_error(
+          "Invalid kernel type. QNNPACK convolution run failed.");
```
nit: message

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -1,11 +1,75 @@
+      &context->quantization_params);
+};
+
+struct QnnpackDeleter {
```
can we move this struct to a common header now that's used in 2 places?

z-a-f replied:
I would love to. The problem is that the qnnpack library is compiled using C11, which doesn't support operator overloading. The `cc` files are compiled using CXX14, which does support it. That's why it is OK to keep the deleter here, but not in the common. I will look into it, because I agree that factoring it out is a good idea -- just need to find a way

z-a-f replied:
I will send a separate PR for this -- it looks like it is a little more involved.

supriyar replied:
Please add a TODO or file an issue to track it

kimishpatel replied:
This might sound like a rant, but let's prioritize deduplication of code. Even the current qnnpack state seems little unmaintainable.

z-a-f replied:
> Please add a TODO or file an issue to track it

Wil  send a separate PR -- we need to discuss what parts of qnnpack should be made available to PT

z-a-f replied:
Moved to the refactoring PR

supriyar reviewed (changes requested):
Left some comments. Can you add test for this API in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/test/deconvolution-operator-tester.h?

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -19,7 +83,242 @@ enum pytorch_qnnp_status qnnpackDeConv(
+  // Check all invalid parameters
+  const size_t kernel_width = deconv_p.kernel_dims[0];
+  const size_t kernel_height = deconv_p.kernel_dims[1];
+  if (kernel_width == 0 || kernel_height == 0) {
```
BTW, why are we doing these checks here? Can this not to be done in deconv_param_t constructor. Conv op does that so we should probably maintain the convention. Even better if we can just use conv_params_t if possible.

supriyar replied:
Yeah I think @z-a-f  mentioned this as a todo in the previous PR. I think we can merge them

z-a-f replied:
This check is done in the params. I can remove it from here

supriyar commented:
@z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

kimishpatel commented:
Can we add test for both inside QNNPACK that can compile for armv7/v8 and python test much like other ones you have?

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -0,0 +1,292 @@
+  const pytorch_q8conv_ukernel_function ukernel;
+};
+
+static void compute_q8conv(
```
There is already a compute_q8conv in conv-run.cc. Lets use that if possible, please?

z-a-f replied:
factored out in #37497

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -0,0 +1,292 @@
+  }
+};
+
+enum pytorch_qnnp_status qnnpackDeConv(
```
I think this and qnnpackConv probably can merge as well with appropriate refactoring, or atleast overlap can be minimized. We can put this under a common umbrella of conv/deconv merge and have a PR to fix that.

z-a-f replied:
These two cannot be merged -- they are two different kernels. I can refactor stuff, but refactoring will be a part of future PRs.

z-a-f replied:
I agree with the overlap -- but I don't think this can be merged with the conv -- these are two different ops, and should live in their own classes.

kimishpatel replied:
I need to think through this. If we have a way to share as much code as possible that would be good. On mobile side this is also an app size issue. From clean code perspective I understand. We shouldn't merge things and make a bloated structures that house unrelated components. For now this is fine.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -0,0 +1,292 @@
+      deconvolution, output_tile_size, tiled_output_size);
+
+  // Run the kernel
+  switch (deconvolution->ukernel_type) {
```
No need for switch here. Only single case.

kimishpatel reviewed (changes requested):
Left some comments with requested changes.

z-a-f commented:
> @z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

I am ironing out some stuff related to the testing. I wanted to add the test in this PR (for future bisecting purposes), but because the prepacking is outside of this PR, I might send the tests separately. I will try to rearrange so that the prepack would be before this -- lemme try

z-a-f commented:
> Can we add test for both inside QNNPACK that can compile for armv7/v8 and python test much like other ones you have?

Python tests cannot use this kernels directly until the python wrappers are written. Those will come later when the python wrappers are written

Commit:
Update on "[quant] QNNPACK deconv"


Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv"


Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv"


Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

z-a-f commented:
> @z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

Done

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/test/deconvolution-operator-tester.h:
```diff
@@ -494,60 +500,108 @@ class DeconvolutionOperatorTester {
+        }
+        break;
+
+        case Mode::Runtime:
```
I think you also need to call the tests with Runtime mode

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:
```diff
@@ -429,7 +430,7 @@ if(PYTORCH_QNNPACK_BUILD_TESTS)
     CXX_EXTENSIONS NO)
   target_include_directories(deconvolution-test PRIVATE src test)
-  target_link_libraries(deconvolution-test PRIVATE pytorch_qnnpack cpuinfo gtest gtest_main)
+  target_link_libraries(deconvolution-test PRIVATE pytorch_qnnpack cpuinfo fp16 gtest gtest_main)
```
Why do we need fp16 library for this?

z-a-f replied:
FP16 is required for the QNNPACK requantization

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc:
```diff
@@ -12,102 +12,93 @@
 #include "convolution-operator-tester.h"
 
-TEST(CONVOLUTION_OP, zero_batch) {
+_STATIC_AND_RUNTIME_TEST(CONVOLUTION_OP, zero_batch,
```
So we're enabling both static and runtime mode for all conv and deconv tests. Do they all pass? I recall you mentioning some of them failed

z-a-f replied:
The same that fail in the regular convolution fail in the transpose as well. I think this is not specific to convolution. Currently, I am enabling the same tests that are enabled from the conv

kimishpatel replied:
Are you sure? Check this on master? On master I think all convolution tests are passing. If it is not failing on master it should not be failing here.

supriyar reviewed (approved):
Looks good, left some comments in the review.

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK deconv kernel and tests"




Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

[ghstack-poisoned]
